geode-core/src/distributedTest/java/org/apache/geode/cache/query/dunit/ResourceManagerWithQueryMonitorDUnitTest.java

```diff
@@ -1273,73 +1037,123 @@ private boolean isExceptionDueToTimeout(QueryException e) {
     String message = e.getMessage();
     // -1 needs to be matched due to client/server set up, BaseCommand uses the
     // MAX_QUERY_EXECUTION_TIME and not the testMaxQueryExecutionTime
+
     return (message.contains("The QueryMonitor thread may be sleeping longer than")
         || message.contains("Query execution canceled after exceeding max execution time")
         || message.contains(
             String.format("Query execution canceled after exceeding max execution time %sms.",
                 -1)));
   }
 
-  private DefaultQuery.TestHook getPauseHook() {
-    return new PauseTestHook();
+  private DefaultQuery.TestHook getPauseHook(boolean hitCriticalThreshold,
+      VM vmToCallback) {
+    return new PauseTestHook(hitCriticalThreshold, vmToCallback);
   }
 
-  private DefaultQuery.TestHook getCancelDuringGatherHook() {
-    return new CancelDuringGatherHook();
+  private DefaultQuery.TestHook getCancelDuringGatherHook(VM vmToCallback) {
+    return new CancelDuringGatherHook(true, vmToCallback);
   }
 
   private DefaultQuery.TestHook getCancelDuringAddResultsHook() {
     return new CancelDuringAddResultsHook();
   }
 
-  private static class PauseTestHook implements DefaultQuery.TestHook {
-    private CountDownLatch latch = new CountDownLatch(1);
-    public boolean rejectedObjects = false;
+  static class RejectedObjectsInterface {
+    boolean rejectedObjects = false;
+  }
+
+  private class PauseTestHook extends RejectedObjectsInterface implements DefaultQuery.TestHook {
+    private final CountDownLatch latch = new CountDownLatch(1);
+
+    boolean hitCriticalThreshold;
+    AtomicBoolean hitOnce = new AtomicBoolean(false);
+    VM callbackVM;
+
+    PauseTestHook(boolean hitCriticalThreshold, VM vmToCallback) {
+      super();
+      this.hitCriticalThreshold = hitCriticalThreshold;
+      callbackVM = vmToCallback;
+    }
 
     @Override
     public void doTestHook(final SPOTS spot, final DefaultQuery _ignored,
         final ExecutionContext executionContext) {
       switch (spot) {
         case BEFORE_QUERY_EXECUTION:
           try {
-            if (!latch.await(8, TimeUnit.SECONDS)) {
-              fail("query was never unlatched");
+            if (!latch.await(8, SECONDS)) {
```

Comment:
I recommend changing the await to use the GeodeAwaitility.getTimeout().

Reply:
Done.